base/condition.jl

```diff
@@ -126,19 +126,56 @@ proceeding.
 function wait end
 
 """
-    wait(c::GenericCondition; first::Bool=false)
+    wait(c::GenericCondition; first::Bool=false, timeout::Real=0.0)
 
 Wait for [`notify`](@ref) on `c` and return the `val` parameter passed to `notify`.
 
 If the keyword `first` is set to `true`, the waiter will be put _first_
 in line to wake up on `notify`. Otherwise, `wait` has first-in-first-out (FIFO) behavior.
+
+If `timeout` is specified, cancel the `wait` when it expires and throw a
+`TimeoutError` to the waiting task. The minimum value for `timeout`
+is 0.001 seconds, i.e. 1 millisecond.
 """
-function wait(c::GenericCondition; first::Bool=false)
+function wait(c::GenericCondition; first::Bool=false, timeout::Real=0.0)
+    timeout == 0.0 || timeout ≥ 1e-3 || throw(ArgumentError("timeout must be ≥ 1 millisecond"))
+
     ct = current_task()
     _wait2(c, ct, first)
     token = unlockall(c.lock)
+
+    timer::Union{Timer, Nothing} = nothing
+    if timeout > 0.0
+        timer = Timer(timeout)
+        # start a task to wait on the timer
+        t = Task() do
+            try
+                wait(timer)
+            catch e
+                # if the timer was closed, the waiting task has been scheduled; do nothing
+                e isa EOFError && return
+            end
+            dosched = false
+            lock(c.lock)
+            # Confirm that the waiting task is still in the wait queue and remove it. If
+            # the task is not in the wait queue, it must have been notified already so we
+            # don't do anything here.
```

Comment:
This appears to introduce a data race though, so we cannot merge this

Reply:
How's that? We're locking the condition variable here.

Reply:
This Timer runs concurrently with the return from `wait`, so by the time this code runs, you might have just corrupted some arbitrary subsequent wait

Reply:
Ah, okay. There's an ABA problem. Let me see if I can find a solution for that.

But the waiting task is only scheduled with a `TimeoutError` if it was in this condition's wait queue, so I'm not sure I understand your "or" case here -- the only subsequent `wait` that could get blown up is a `wait` on the same condition, which is the same ABA problem?

Reply:
It could been in the waitq, then removed before you got around to scheduling it, or vice versa with some other thread scheduling before it got around to removing it from the queue. Those codes are running on other threads, so it could be concurrent. There is potentially no guarantee that you can safely mutate this data-structure concurrently on two threads (https://github.com/JuliaLang/julia/pull/55542)

Reply:
Pushed a fix for the ABA problem that relies on happens-before -- if the waiter was scheduled, it sets `waiter_left` before returning. It can only re-enter the condition's wait queue by another call to `wait`, for which it must acquire the lock.

We acquire the condition's lock before checking `waiter_left` and for the task's presence in the wait queue. If the task is present, it can only be because it has not been scheduled, because if it was scheduled, it would have set `waiter_left` before re-entering the wait queue.

I think the combination of the lock and the atomic assure there is no ABA problem.

Reply:
>It could been in the waitq, then removed before you got around to scheduling it

We acquire the lock, confirm that the waiter did not leave and remove it from the wait queue before scheduling it. If it was not in the wait queue, we do not schedule it and this decision is made while holding the lock.

>some other thread scheduling before it got around to removing it from the queue

If the task is scheduled by `notify`, then it is removed from the condition's wait queue before it is scheduled, which is done while holding the condition's lock. If it is not in the wait queue, then we do not schedule it.